[awdse22] iP

ngkhengyang opened the pull request:
# MrYapper Task Management Bot

> "I actually don't yap a lot despite my name... I think(?)" - MrYapper

Helps you keep track of all the task and list out whichever tasks you have!
- Text-based
- Simple to use, does **not** yap unecessarily
- ~~Very~~ SUPER efficient!!

How do you use the bot?
1. Clone from [this repository](https://github.com/awdse22/ip)
2. Compile the java files
3. Run the MrYapper class

...and that's it, it's so simple 😱 

## Features

- [x] Add different kinds of task
- [x] Display and search tasks
- [x] Mark tasks as done or not done
- [ ] GUI Interface 
> [!NOTE]
> Chatbot is still under development and more features may be implemented in the future

Here is the preview of the code which includes the main program:
```java
public void run() {
    boolean conversationIsOngoing = false;
    if (tasks != null && storageManager != null) {
        ui.showGreeting();
        conversationIsOngoing = true;
    }

    while (conversationIsOngoing) {
        try {
            String userInput = ui.readCommand();
            ui.showLine();
            Command c = Parser.parse(userInput);
            conversationIsOngoing = !c.execute(tasks, ui, storageManager);
        } catch (IllegalTaskException | InvalidSyntaxException | IllegalArgumentException e) {
            ui.send(e.getMessage());
        }
        ui.showLine();
    }
    ui.close();
}

public static void main(String[] args) {
    new MrYapper(TASK_DATA_PATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ngkhengyang committed:
Renamed main class

ngkhengyang committed:
Changed greeting message

ngkhengyang committed:
Add echo function and bye command

ngkhengyang committed:
Re-organise code

* Rename variables to follow java coding standard
* Separate logical units and functions for enhanced readability

ngkhengyang committed:
Create Task class and function to add Task

ngkhengyang committed:
Add function to display task list

ngkhengyang committed:
Modify format of response when adding task

ngkhengyang committed:
Add function to mark tasks as done

ngkhengyang committed:
Make list display whether task is done

ngkhengyang committed:
Add function to unmark task and change main to use switch statements

ngkhengyang committed:
Create Todo task class

ngkhengyang committed:
Create deadline task type and class

ngkhengyang committed:
Add Event task type

ngkhengyang committed:
Add InvalidTaskException for invalid inputs on task creation

ngkhengyang committed:
Add support for deleting tasks from the list

ngkhengyang committed:
Add semi-automated text UI testing

ngkhengyang committed:
Add function to check for data file and create one if needed

ngkhengyang committed:
Make the chatbot process the data file from hard disk if any

ngkhengyang committed:
Make adding tasks write into the data file

ngkhengyang committed:
Rename parseTaskData function to readTaskData

ngkhengyang committed:
Make the chat bot save all changes to task list when closing

ngkhengyang committed:
Merge branch 'branch-level-7'

ngkhengyang committed:
Add parsing of date and time from text inputs

ngkhengyang committed:
Merge branch 'branch-level-8'

ngkhengyang committed:
Extract task list to be contained in TaskList class

ngkhengyang committed:
Extract a StorageManager class to manage loading and saving data

ngkhengyang committed:
Extract classes Command, Parser and Ui and refactor code

ngkhengyang committed:
Merge branch 'branch-A-MoreOOP'

ngkhengyang committed:
Divide classes into packages

ngkhengyang committed:
Merge branch 'branch-A-Packages'

ngkhengyang committed:
Merge branch 'add-gradle-support' of https://github.com/awdse22/ip

ngkhengyang committed:
Set up gradle and allow building and running using gradle

ngkhengyang committed:
Add JUnit test case directory and dummy tests

ngkhengyang committed:
Add unit test cases for Todo, Event and Deadline

ngkhengyang committed:
Refactor display of messages and TaskList saveToStorage method

ngkhengyang committed:
Add unit test cases for TaskList class

ngkhengyang committed:
Add unit test cases for Parser class

ngkhengyang committed:
Move TaskStub class into test case directory

ngkhengyang committed:
Merge branch 'branch-A-JUnit'

ngkhengyang committed:
Change base name of jar file when building with shadowJar

ngkhengyang committed:
Merge branch 'branch-A-Jar'

yxtan02 commented on src/main/java/mryapper/MrYapper.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public void run() {
+        boolean conversationIsOngoing = false;
```
Would it be better for the boolean variable conversationIsOngoing to be renamed to isConversationOngoing? According to the coding standards, boolean variables should be prefixed with is, has, was as much as possible.

yxtan02 commented on src/main/java/mryapper/MrYapper.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.io.IOException;
+
+public class MrYapper {
```
Would it be better to add Javadoc header comment for this class? According to the coding standards, header comments should be added for all public classes/methods.

yxtan02 commented on src/main/java/mryapper/MrYapper.java:
```diff
@@ -0,0 +1,54 @@
+
+public class MrYapper {
+
+    private static final String TASK_DATA_PATH = "src/data/tasks.txt";
```
Good job for naming your constant in all uppercase and using snake case!

yxtan02 commented on src/main/java/mryapper/exception/IllegalTaskException.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public String getMessage() {
+        String errorMessage = baseMessage + "\n";
+        switch (taskType) {
+            case "todo":
+                errorMessage += " To add a todo task, I need the task description\n"
+                        + " e.g. todo read CS2103T notes";
+                break;
+            case "deadline":
+                errorMessage += " I need one description and deadline using \"/by\"\n"
+                        + " e.g. deadline CS2103T project /by Dec 31st";
+                break;
+            case "event":
+                errorMessage += " I need one description, start and end time using \"/from\" and \"/to\"\n"
+                        + " e.g. event project meeting /from Mon 2pm /to 4pm";
+        }
```
Should the switch statement is reformatted to follow the coding standard (no indentation for case clauses) ?

yxtan02 commented on src/main/java/mryapper/command/CommandList.java:
```diff
@@ -0,0 +1,102 @@
+import mryapper.task.Todo;
+import mryapper.task.Event;
+
+public class CommandList {
```
Would it be better to split this class into multiple subclasses for each specific command?

yxtan02 reviewed (commented):
Looks good to me. Just a few nits to fix.

ngkhengyang committed:
Add JavaDoc for majority of the classes

ngkhengyang committed:
Fix various method names and use of constants

ngkhengyang committed:
Add a command to find task based on given keywords

ngkhengyang committed:
Merge branch 'branch-A-JavaDoc'

ngkhengyang committed:
Merge branch 'branch-A-CodingStandard'

ngkhengyang committed:
Merge branch 'branch-level-9'

NgZiXin commented on src/main/java/mryapper/task/Task.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    /**
+<<<<<<< HEAD
```
Rmb to resolve your conflicts

NgZiXin commented on src/main/java/mryapper/task/Task.java:
```diff
@@ -0,0 +1,99 @@
+     * @param searchInput The search input from the user.
+     * @return Whether all keywords are contained in the task description.
+     */
+    public boolean hasKeywords(String searchInput) {
```
Good naming for method returning boolean.

NgZiXin commented on src/main/java/mryapper/task/Event.java:
```diff
@@ -0,0 +1,33 @@
+ * An event task that occurs within the specified time.
+ */
+public class Event extends Task {
+    private final String eventStart;
```
Can just name as start and end (But NO violation here, just thought it sounds more intuitive).

NgZiXin commented on src/main/java/mryapper/storagemanager/StorageManager.java:
```diff
@@ -0,0 +1,110 @@
+                throw new InvalidFileDataException(taskData);
+            }
+
+            boolean taskIsDone;
```
Name as isTaskDone instead.

NgZiXin commented on src/main/java/mryapper/task/Task.java:
```diff
@@ -0,0 +1,99 @@
+
+    // gets the string data to be written in the data file
+>>>>>>> branch-level-9
+    public abstract String getDataString();
```
Remember to have have proper line spacing.

NgZiXin replied:
Nitpicking here, but can consider getFormattedString for clarity.

NgZiXin reviewed (commented):
LGTM. Just minor comments on your styles and choice of names. Also, please resolve your merge conflicts and run automated tests to ensure the code works.

ngkhengyang committed:
Fix various coding style violations and add more header comments

ngkhengyang committed:
Add basic GUI

ngkhengyang committed:
Implement chatbot functions into GUI

ngkhengyang committed:
Merge branch 'branch-level-10'

ngkhengyang committed:
Add assert feature

ngkhengyang committed:
Extract methods and refactor code

* Some methods have very long method bodies.
* The Ui class does not serve much purpose with GUI,
it has many unused methods.
* Switch statements have missing default cases.

Long method bodies and redundant code makes code less
readable and more difficult to catch bugs.

Let's,
* Remove the Ui class to extract only relevant code.
* Extract methods to follow Single Level of Abstraction
Principle.

Doing so breaks longer methods down to smaller logical
blocks and gets rid of irrelevant code for better
readability and ease in debugging.

ngkhengyang committed:
Make the application exit upon error loading data file

If there is an error creating the data file or the data
file is not of the correct format, the chatbot proceeds
to run as usual.

Having the chatbot run even without a proper data file
can introduce bugs during runtime and worsen user
experience.

Let's make the chatbot verify the data file upon
initialization and exit if there is no proper data
file.

This allows the user to know that there is an issue
regarding the data file and should check their hard
disk or delete the old data file instead of encountering
bugs when using the chatbot.

ngkhengyang committed:
Merge pull request #1 from awdse22/branch-A-Assertions

Add assert feature

ngkhengyang committed:
Merge branch 'master' into branch-A-CodeQuality

ngkhengyang committed:
Merge pull request #2 from awdse22/branch-A-CodeQuality

Branch a code quality

ngkhengyang committed:
Extract all the commands into different classes

ngkhengyang committed:
Modify how how InvalidSyntaxException is displayed to users

ngkhengyang committed:
Rename StorageManager to Storage

ngkhengyang committed:
Implement feature to edit tasks

ngkhengyang committed:
Merge branch 'branch-C-Update'

ngkhengyang committed:
Extract method to check for positive task number

ngkhengyang committed:
Update Parser JUnit test cases and some error messages in Parser class

ngkhengyang committed:
Update remaining JUnit test cases and added header comments

ngkhengyang committed:
Fix various coding standard violations and header comments

ngkhengyang committed:
Customize the GUI appearance

ngkhengyang committed:
Fix bug with DateTimeParser not functioning as intended

ngkhengyang committed:
Add user guide and apply a few minor bug fixes

ngkhengyang committed:
Fix bug with jar file being unable to create data file

ngkhengyang committed:
Fix marking and unmarking task not saving in data file
